rectools/models/nn/transformer_data_preparator.py

```diff
@@ -0,0 +1,375 @@
+        sessions = (
+            interactions.sort_values(Columns.Datetime, kind="stable")
+            .groupby(Columns.User, sort=sort_users)[[Columns.Item, Columns.Weight]]
+            .agg(list)
+        )
+        sessions, weights = (
+            sessions[Columns.Item].to_list(),
+            sessions[Columns.Weight].to_list(),
+        )
+
+        return cls(sessions=sessions, weights=weights)
+
+
+class TransformerDataPreparatorBase:
+    """
+    Base class for data preparator. To change train/recommend dataset processing, train/recommend dataloaders inherit
+    from this class and pass your custom data preparator to your model parameters.
+
+    Parameters
+    ----------
+    session_max_len : int
+        Maximum length of user sequence.
+    batch_size : int
+        How many samples per batch to load.
+    dataloader_num_workers : int
+        Number of loader worker processes.
+    item_extra_tokens : Sequence(Hashable)
+        Which element to use for sequence padding.
+    shuffle_train : bool, default True
+        If ``True``, reshuffles data at each epoch.
+    train_min_user_interactions : int, default 2
+        Minimum length of user sequence. Cannot be less than 2.
+    get_val_mask_func : Callable, default None
+        Function to get validation mask.
+    """
+
+    train_session_max_len_addition: int = 0
+
+    item_extra_tokens: tp.Sequence[Hashable] = (PADDING_VALUE,)
+
+    def __init__(
+        self,
+        session_max_len: int,
+        batch_size: int,
+        dataloader_num_workers: int,
+        shuffle_train: bool = True,
+        train_min_user_interactions: int = 2,
+        n_negatives: tp.Optional[int] = None,
+        get_val_mask_func: tp.Optional[tp.Callable] = None,
+        **kwargs: tp.Any,
+    ) -> None:
+        self.item_id_map: IdMap
+        self.extra_token_ids: tp.Dict
+        self.train_dataset: Dataset
+        self.val_interactions: tp.Optional[pd.DataFrame] = None
+        self.session_max_len = session_max_len
+        self.n_negatives = n_negatives
+        self.batch_size = batch_size
+        self.dataloader_num_workers = dataloader_num_workers
+        self.train_min_user_interactions = train_min_user_interactions
+        self.shuffle_train = shuffle_train
+        self.get_val_mask_func = get_val_mask_func
+
+    def get_known_items_sorted_internal_ids(self) -> np.ndarray:
+        """Return internal item ids from processed dataset in sorted order."""
+        return self.item_id_map.get_sorted_internal()[self.n_item_extra_tokens :]
+
+    def get_known_item_ids(self) -> np.ndarray:
+        """Return external item ids from processed dataset in sorted order."""
+        return self.item_id_map.get_external_sorted_by_internal()[self.n_item_extra_tokens :]
+
+    @property
+    def n_item_extra_tokens(self) -> int:
+        """Return number of padding elements"""
+        return len(self.item_extra_tokens)
+
+    @staticmethod
+    def _process_features_for_id_map(
```

Comment:
We are now keeping all of the features that were present in dataset, both Dense and Spare variants